### PR TITLE
feat(server): add logits watermark

### PR DESCRIPTION
**Pull request by OlivierDehaene**

### CODE REVIEW

File: server/text_generation/utils/watermark.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import torch
+from transformers import LogitsProcessor
+
+
+class WatermarkLogitsProcessor(LogitsProcessor):
+    def __init__(
+        self,
+        vocab_size: int,
+        gamma: float = 0.5,
```
Comment (OlivierDehaene) [inline server/text_generation/utils/watermark.py:25]: All these parameters need to be fixed for the detector to work right?

File: server/text_generation/utils/watermark.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    def _bias_greenlist_logits(
+        scores: torch.Tensor, greenlist_mask: torch.Tensor, greenlist_bias: float
+    ) -> torch.Tensor:
+        scores[greenlist_mask] = scores[greenlist_mask] + greenlist_bias
+        return scores
+
+    def __call__(
+        self, input_ids: torch.LongTensor, scores: torch.FloatTensor
+    ) -> torch.FloatTensor:
+        assert len(input_ids) == 1
```
[the remark this answers is not in the document]
Comment (jwkirchenbauer) [reply server/text_generation/utils/watermark.py:74]: to clarify - this is statement about the requirements on your side? not a question about _whether_ or not the watermark can be applied to the batch?

The reason I ask is that it's actually true that generally I've done most evaluation work with _input_ bsz of 1, with the caveat that beam search uses the batch dim itself to update hypotheses, so it does work in the batched setting. It's just not the most efficient because the seeding/rng operation itself isn't trvially parallelizable to my knowledge.

File: server/text_generation/utils/watermark.py (hunk 1 above)
Comment (jwkirchenbauer) [reply server/text_generation/utils/watermark.py:25]: Yes, or rather _they need to match at detection time_. The core assumption is realized by the class structure. 

The base class logic, parametrized by kwargs, is the contract between the processor (watermark-_er_) and the detector. If they don't match, then the detector will (in expectation) return False/unwatermarked, since the pseudorandom procedure won't "line up" and thus the outputs will read as unwatermarked.

File: server/text_generation/utils/watermark.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (OlivierDehaene) [reply server/text_generation/utils/watermark.py:74]: No it was just a remark to help your review :) 
We are always in the batch size == 1 setting here as each request is warped individually.

[the remark this answers is not in the document]
Comment (jwkirchenbauer) [reply server/text_generation/utils/watermark.py:74]: A follow-up: does this mean that beam search is not supported by text-generation-inference? 
This is totally okay if so, but if it is supported, I take it that the beam management is handled separately/differently than in the normal generation pipeline? I sort of assume that each beam must be treated as an indiv request, and is warped individually as a consequence. Just want to make sure it's absolutely okay to lose that batch loop from the orig code.

File: server/text_generation/utils/tokens.py [hunk 3]
```diff
@@ -47,6 +50,11 @@ def __init__(
         # the following idea is largely copied from this PR: https://github.com/huggingface/transformers/pull/5420/files
         # all samplers can be found in `generation_utils_samplers.py`
         sampling = do_sample
+
+        if watermark:
+            warpers.append(WatermarkLogitsProcessor(vocab_size, device=device))
```
Comment (jwkirchenbauer) [inline server/text_generation/utils/tokens.py:55]: I don't see kwargs (like `delta`, `gamma`) being passed at construction time here. Does this mean that a fixed version of the watermark will be instantiated for all generations?

**Comment by jwkirchenbauer** [comment]
Just a few questions:
1. Confirming that the logic for changing the "seeding_scheme" has been removed intentionally - this makes sense, since it was a forward looking demonstration through implementation that other schemes, as discussed in the paper, are possible.
2. Will it be possible to call generate with different watermark parameters for each generation (similar to different temperatures or repetition penalties)? Or will the watermark parameters be fixed after the endpoint is spun up?  I made a more targeted comment on the processor construction line.

File: server/text_generation/utils/watermark.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (OlivierDehaene) [reply server/text_generation/utils/watermark.py:74]: There is no beam search in text-generation-inference.

File: server/text_generation/utils/tokens.py (hunk 3 above)
Comment (OlivierDehaene) [reply server/text_generation/utils/tokens.py:55]: Yes indeed.

**Comment by jwkirchenbauer** [comment]
> Well yes we can add the options to do that. What would be the use case though? It will completely bypass the watermark if request can set up their own parameters, right? However I can add the possibility to set these parameters when we first start up the service.

Sounds good. So, you're totally right, in the "real deployment" usecase the watermark parameters would be controlled and fixed by the model provider (the detector itself would likely be run by them too). The ability to produce generations with various parameters somewhat on-demand is more of a capability desired for testing and evaluation purposes on the research side of things. I'm happy to spin with fixed parameters for now, so I'll complete the review. 

Let me know what the next step is going to be/how I can help bring up and properly utilize a widget that applies the new processor!

**Review by jwkirchenbauer — APPROVED** [review]

**Review by jwkirchenbauer — COMMENTED** [review]
Without understanding the system much at all, I assume there is a way to set the env vars at startup which allows a widget to be spun up with a specific value of gamma and delta. I assume this doesn't allow individual queries to be sent with those specific parameters? (and this is fine/not required)
